### PR TITLE
Computer vision initial services

### PR DESCRIPTION
**Pull request by javiermtorres**
Preliminary implementation for computer vision tasks (object detection, image segmentation and image classification).

### CODE REVIEW

File: encoderfile/proto/image_classification.proto [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+import "proto/token.proto";
+import "proto/metadata.proto";
+
+service ImageClassification {
+  rpc Predict(ImageClassificationRequest) returns (ImageClassificationResponse);
+  rpc GetModelMetadata(encoderfile.metadata.GetModelMetadataRequest) returns (encoderfile.metadata.GetModelMetadataResponse);
+}
+
+message ImageInput {
+  bytes image = 1;
```
Comment (besaleli) [inline encoderfile/proto/image_classification.proto]: nit: having 3 different `ImageInput` feels like a code smell. How would you feel about either/any combination of:

- just throwing the images directly in `ImageClassificationRequest` and doing something like `repeated bytes inputs = 1;`. This would give us parity with `repeated string inputs = 1;` in text services
- Putting it in a separate proto or reusing it
- Condensing all of the protos into one file (would this be so horrible... this may make it easier if anyone is working with Kafka)

File: encoderfile/proto/object_detection.proto [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  ymin int32 = 5;
+  ymax int32 = 6;
+}
+
+message ImageBoundingBoxes {
+  repeated ImageBoundingBox boxes = 1;
+}
+
+message ObjectDetectionResponse {
+  repeated ImageBoundingBoxes boxes_batch = 1;
```
Comment (besaleli) [inline encoderfile/proto/object_detection.proto]: nit: would prefer this just be named `boxes`, but not a hill I need to die on

**Review by besaleli — COMMENTED** [review]
Thoughts!!

File: encoderfile/proto/object_detection.proto (hunk 2 above)
Comment (javiermtorres) [reply encoderfile/proto/object_detection.proto]: The thing is that for each image, we have boxes, but then for a bunch of input images we have _groups_ of boxes.... so I ran out of imagination :-/ Any suggestions here?

File: encoderfile/proto/image_classification.proto (hunk 1 above)
Comment (javiermtorres) [reply encoderfile/proto/image_classification.proto]: I'm now taking image classification as "gold standard" while I work at the impl, and I'll probably retrofit the other two categories from there. So I'll take these comments into consideration when I revisit the proto defs for them 👍

Comment (javiermtorres) [reply encoderfile/proto/image_classification.proto]: (It just felt easier to transcribe the hf pipelines in separate files at the start)

Comment (javiermtorres) [reply encoderfile/proto/image_classification.proto]: (separate proto + imports ftw, imho)

File: encoderfile/proto/object_detection.proto (hunk 2 above)
Comment (javiermtorres) [reply encoderfile/proto/object_detection.proto]: renamed to "box" and "boxes", although the grammatical number agreement gets a bit weird :-/

File: encoderfile-runtime/src/main.rs [hunk 3]
```diff
@@ -30,49 +27,78 @@ async fn main() -> Result<()> {
 }
 
 macro_rules! run_cli {
-    ($model_type:ident, $cli:expr, $config:expr, $session:expr, $tokenizer:expr, $model_config:expr) => {{
+    ($model_type:ident, $cli:expr, $config:expr, $session:expr, $input_state:expr, $task_state:expr) => {{
```
Comment (besaleli) [inline encoderfile-runtime/src/main.rs]: merge in latest changes, we changed `encoderfile-runtime/main.rs` in the gpu update and model type dispatch is now happening elsewhere

File: encoderfile/src/runtime/state.rs [hunk 4]
```diff
@@ -1,32 +1,170 @@
...
+
 #[derive(Debug)]
-pub struct EncoderfileState<T: ModelTypeSpec> {
+pub struct EncoderfileState<T: ModelTypeSpec + InputType + TaskType> {
     pub config: Config,
     pub session: Mutex<Session>,
-    pub tokenizer: TokenizerService,
-    pub model_config: ModelConfig,
+    pub per_model_input_state: T::InputState,
+    pub per_task_state: T::TaskState,
```
Comment (besaleli) [inline encoderfile/src/runtime/state.rs]: naming here unclear. what does per_model and per_task mean? why can it just not be `model_input_state` and `task_state`?

File: encoderfile/src/common/model_type.rs [hunk 5]
```diff
@@ -1,5 +1,9 @@
 macro_rules! model_type {
     [ $( $x:ident ),* $(,)? ] => {
+        pub trait ModelTypeSpec: Send + Sync + Clone + std::fmt::Debug + 'static {
```
Comment (besaleli) [inline encoderfile/src/common/model_type.rs]: why are we moving this inside the macro?

File: encoderfile/src/dev_utils/mod.rs [hunk 6]
```diff
@@ -1,20 +1,31 @@
...
+        TaskType,
+    },
 };
 use ort::session::Session;
 use parking_lot::Mutex;
 use std::str::FromStr;
 use std::{fs::File, io::BufReader};
 
 const EMBEDDING_DIR: &str = "../models/embedding";
+// CHECK sentence embedding????
```
Comment (besaleli) [inline encoderfile/src/dev_utils/mod.rs]: embedding and sentence embedding can use the same model

File: encoderfile/src/format/assets/kind.rs [hunk 7]
```diff
@@ -50,31 +50,48 @@ impl AssetKind {
...
-                    AssetKind::ModelConfig,
-                    AssetKind::Tokenizer,
-                ]
-            }
-
-            fn optional_assets() -> &'static [AssetKind] {
-                &[AssetKind::Transform]
-            }
-        }
+        impl AssetPolicySpec for crate::common::model_type::$model_type {}
```
Comment (besaleli) [inline encoderfile/src/format/assets/kind.rs:98]: if we are moving all of the logic out of the trait, what is the point in having the trait? we might as well just make it a function

File: encoderfile/src/inference/image_classification.rs [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+) -> Result<Vec<Vec<ImageLabelScore>>, ApiError> {
+    let grouped_images = ort::value::TensorRef::from_array_view(
+        &images)
+        .unwrap()
+        .to_owned();
+    let raw_outputs = crate::run_cv_model!(session, grouped_images)?;
+    println!("Raw outputs: {:?}", raw_outputs.keys().collect::<Vec<_>>());
+    let mut outputs = raw_outputs
+        .get("logits")
+        .expect("Model does not return logits")
```
Comment (besaleli) [inline encoderfile/src/inference/image_classification.rs]: get rid of expects. these should throw errors

File: encoderfile/src/services/image_classification.rs [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+const DEFAULT_FILTER_TYPE: image::imageops::FilterType = image::imageops::FilterType::Triangle;
+
+impl Inference for AppState<model_type::ImageClassification>
+{
+    type Input = ImageClassificationRequest;
+    type Output = ImageClassificationResponse;
+
+    fn inference(&self, request: impl Into<Self::Input>) -> Result<Self::Output, ApiError> {
+        let request = request.into();
+        let rescale_factor = 0.00392156862745098 as f32;
```
Comment (besaleli) [inline encoderfile/src/services/image_classification.rs]: we should make these preprocessing steps into lua bindings. with a `Preprocess` function that's extracted from the transform

File: encoderfile/src/transforms/engine/image_classification.rs [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+        let func = match self.postprocessor() {
+            Some(p) => p,
+            None => return Ok(data),
+        };
+
+        let expected_shape = data.shape().to_owned();
+
+        let tensor = Tensor(data.into_dyn());
+
+        let result = func
```
Comment (besaleli) [inline encoderfile/src/transforms/engine/image_classification.rs:20]: slay

File: encoderfile/src/transport/http/example.md [hunk 11]
```diff
@@ -0,0 +1,316 @@
+# Multipart OpenAPI Service Example
```
Comment (besaleli) [inline encoderfile/src/transport/http/example.md:1]: @angpt looping you in here. we should add this into the docs when we release new version

File: encoderfile-runtime/src/main.rs [hunk 12]
```diff
@@ -30,49 +27,78 @@ async fn main() -> Result<()> {
...
         let state = Arc::new(EncoderfileState::<$model_type>::new(
             $config,
             $session,
-            $tokenizer,
-            $model_config,
+            $input_state,
```
Comment (besaleli) [inline encoderfile-runtime/src/main.rs]: the model type dispatch got moved to `encoderfile` from `encoderfile-runtime`. i would merge conflicts from main bc there are big code diffs

File: encoderfile/src/services/image_classification.rs (hunk 9 above)
Comment (besaleli) [inline encoderfile/src/services/image_classification.rs]: the rescaling etc. is going into a lua library, yes? my intuition was that users would put preprocessing things in a `Preprocess` function

Comment (javiermtorres) [reply encoderfile/src/services/image_classification.rs]: Yep. I wanted to make it work first, but totally agreed. Let's see if we can get something closer to what one would expect in an hf pipeline.
BTW I'm not considering b/w (1 channel) images, for example, right now.

File: encoderfile/src/transforms/engine/image_classification.rs (hunk 10 above)
Comment (javiermtorres) [reply encoderfile/src/transforms/engine/image_classification.rs:20]: I'll revisit this, I think I just copied whatever it was already there without much consideration 😂
I guess we can do the same for text logits in this case, but it will get more complicated for object detection and image segmentation. But shape checks will fix everything 😉

File: encoderfile/src/runtime/state.rs (hunk 4 above)
Comment (javiermtorres) [reply encoderfile/src/runtime/state.rs]: I thought the "per_" would convey the meaning that this is an input state / task state (a config really) that depends on what input / task are used in your model. But I guess we can get rid of the per_ thing if it causes confusion.
At some point the state and config concepts need to be separated, but not today :)

File: encoderfile/src/common/model_type.rs (hunk 5 above)
Comment (javiermtorres) [reply encoderfile/src/common/model_type.rs]: I think this was a leftover from some previous restructuring. I'll take it back out.

Comment (javiermtorres) [reply encoderfile/src/common/model_type.rs]: Moved back out.

File: encoderfile/src/format/assets/kind.rs (hunk 7 above)
Comment (javiermtorres) [reply encoderfile/src/format/assets/kind.rs:98]: Quite legit.... I guess we could just have a `fn required_assets(spec: impl InputType + TaskType)`. I doubt we could gain something by moving this to type space 😕

Also, the new InputType and TaskType traits (or vals in a fun) should be enough to compose the list of assets, instead of matching the tuples, right?

File: encoderfile/proto/image_classification.proto [hunk 13]
```diff
@@ -0,0 +1,21 @@
+syntax = "proto3";
```
Comment (besaleli) [inline encoderfile/proto/image_classification.proto:1]: one day we're gonna have to merge all of these so they're actually usable by someone streaming data into kafka for example 😩

File: encoderfile/src/builder/image_preprocessor.rs [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+// models on a case-by-case basis as they come in.
+
+use crate::format::assets::{AssetKind, AssetSource, PlannedAsset};
+use anyhow::Result;
+
+use super::config::EncoderfileConfig;
+use crate::runtime::ImagePreprocessing;
+
+pub fn validate_image_preprocessor<'a>(
+    efconfig: &'a EncoderfileConfig,
```
Comment (besaleli) [inline encoderfile/src/builder/image_preprocessor.rs]: just `config` or `encoderfile_config`? we may want some naming conventions soon

File: encoderfile/src/builder/config.rs [hunk 15]
```diff
@@ -24,7 +24,7 @@ pub struct BuildConfig {
     pub encoderfile: EncoderfileConfig,
 }
 
-pub const DEFAULT_VERSION: &str = "0.1.0";
+pub const DEFAULT_VERSION: &str = "0.2.0";
```
Comment (besaleli) [inline encoderfile/src/builder/config.rs:27]: if we're gonna change this, we need backwards compatibility logic

File: encoderfile/src/services/image_classification.rs [hunk 16]
```diff
@@ -0,0 +1,253 @@
...
+        let images_array = ndarray::stack(
+            Axis(0),
+            &images.iter().map(|x| x.view()).collect::<Vec<_>>(),
+        )
+        .unwrap()
+        .into_dimensionality::<Ix4>()
+        .unwrap();
+
+        // TODO make parallel???
+        // TODO _maybe
```
Comment (besaleli) [inline encoderfile/src/services/image_classification.rs]: I wish these comments were less cryptic

File: encoderfile/src/services/image_classification.rs [hunk 17]
```diff
@@ -0,0 +1,253 @@
...
+                .map(|labels| ImageClassificationResult {
+                    labels: labels.clone(),
+                })
+                .collect(),
+            metadata: request.metadata,
+        })
+
+        /*
+
+                let rescale_factor = self.model_input_state.preprocessing.rescale_factor.ok_or(ApiError::InternalError("missing rescale factor"))?;
```
Comment (besaleli) [inline encoderfile/src/services/image_classification.rs]: ?????

File: encoderfile/src/services/image_classification.rs [hunk 18]
```diff
@@ -0,0 +1,253 @@
...
+            return img:resize(224,224):to_array(3)
+        end
+        "##
+        .to_string();
+
+        let engine =
+            ImageClassificationTransform::new(DEFAULT_LIBS.to_vec(), Some(postprocess_code))
+                .expect("Failed to create engine");
+
+        let num_channels = self.model_input_state.config.num_channels as usize;
```
Comment (besaleli) [inline encoderfile/src/services/image_classification.rs:40]: ~~If I'm understanding correctly, we are now making users create an entirely separate config to provide values that *we're* deciding they need for postprocessing. I was hoping that users could define these in the lua transform and have more control over the exact steps they want to use. This is the point of having the lua transform—so that users don't get config fatigue~~

Seeing now that `preprocessor_config.json` is an asset for hf image classification models. Might be helpful to include a comment. Have we checked different models to see if the schema gets crazy? They don't like to give us authoritation JSON schemas :')

File: encoderfile/src/transport/http/example.md [hunk 19]
```diff
@@ -0,0 +1,316 @@
...
+## Example 1: cURL with Two Image Files
+
+```bash
+curl -X POST http://localhost:8080/predict/multipart \
+  -F "payload={\"model_version\": \"1.0\", \"threshold\": 0.8}" \
+  -F "files=@/path/to/image1.png" \
+  -F "files=@/path/to/image2.jpg"
+```
+
+### Request Body (multipart/form-data)
```
Comment (besaleli) [inline encoderfile/src/transport/http/example.md:19]: nice

File: encoderfile/src/transport/cli.rs [hunk 20]
```diff
@@ -110,9 +110,9 @@ pub enum Commands {
 }
 
 impl Commands {
-    pub async fn execute<'a, R: Read + Seek>(
+    pub async fn execute<'loader, R: Read + Seek>(
```
Comment (besaleli) [inline encoderfile/src/transport/cli.rs:113]: I appreciate this rename

File: encoderfile/tests/test_models.rs [hunk 21]
```diff
@@ -153,8 +162,35 @@ fn test_token_classification_inference_with_bad_model() {
...
         &transform,
-        &state.model_config,
+        &state.task_state,
         encodings.clone(),
     )
     .expect("Failed to compute results");
 }
+
+#[test]
+fn test_image_classification_model() {
```
Comment (besaleli) [inline encoderfile/tests/test_models.rs:172]: ?

File: encoderfile/src/builder/image_preprocessor.rs (hunk 14 above)
Comment (javiermtorres) [reply encoderfile/src/builder/image_preprocessor.rs]: config feels ambiguous and encoderfile_config feels too long T_T 
yes to some naming discussion :D In the meantime, I'll switch to plain config.

File: encoderfile/src/builder/config.rs (hunk 15 above)
Comment (javiermtorres) [reply encoderfile/src/builder/config.rs:27]: Hum. The runtime won't need it because the payload is bundled with it anyway, but it matters for the builder. However, both are going to be present for every version, so one just needs to get the builder from the same tag as the runtime. That's how we can get away with backwards compatibility, but maybe it's not too user friendly :-/

File: encoderfile/src/services/image_classification.rs (hunk 16 above)
Comment (javiermtorres) [reply encoderfile/src/services/image_classification.rs]: Hum, erm, totally right 😅😅😅😅😅
Ok, so, in the case of text processing, the preprocessing is not even comparable to inference, and we can do both sequentially. However, for image processing, we could think about overlapping image preprocessing and image inference, choosing a proper batch number instead of just putting everything in one batch. I'd still assume inference is going to be larger anyway, so the savings may not be too big (specially if everything is done via CPU, haw); otoh, this helps introducing queues for some streaming mode of operation later on.

I hope this makes it clearer. I don't think we need to start implementing any of this now, though, but I'll make like a less cryptic comment here 🤣🙈

File: encoderfile/src/services/image_classification.rs (hunk 17 above)
Comment (javiermtorres) [reply encoderfile/src/services/image_classification.rs]: I was keeping this just as a reference (so this whole comment block will just go away). The idea is to do most of the preproc using the Lua script:

```
        let postprocess_code = r##"
        function Preprocess(img)
            return img:resize(224,224):to_array(3)
        end
        "##
```

The part building the mean and std broadcasts should probably be done in Lua as well; I just need to add some more funs to the Array Lua support.

File: encoderfile/src/services/image_classification.rs (hunk 18 above)
Comment (javiermtorres) [reply encoderfile/src/services/image_classification.rs:40]: Sorry!!! Yes, apparently `preprocessor_config.json` is a thing in image models. Yes, as expected, it is largely undocumented (I'm just browsing the pipeline code...) and sometimes it overlaps with `config.json`. Ha. Hahaha.
I have indeed checked a couple of models or three. The entries I use here seem consistently used anywhere else with the same format (and we'll deal with overlapping entries later on), but there are some (like `channels`) that would have a greater impact in our code if we really want to handle them.

Comment (javiermtorres) [reply encoderfile/src/services/image_classification.rs:40]: This reminds me that we need to instruct the user to copy yet another config file, or maybe make simple shell/PowerShell scripts 🤔

File: encoderfile/src/transport/http/example.md (hunk 19 above)
Comment (javiermtorres) [reply encoderfile/src/transport/http/example.md:19]: Hahaha. Wait until it really works 😋 But yes, multipart seemed ok here (and maybe for stdin?)

File: encoderfile/src/transport/cli.rs (hunk 20 above)
Comment (javiermtorres) [reply encoderfile/src/transport/cli.rs:113]: I might change a couple of other lifetime refs so I really understand what they are referring to....

File: encoderfile/tests/test_models.rs (hunk 21 above)
Comment (javiermtorres) [reply encoderfile/tests/test_models.rs:172]: Placeholder, sorry. I wanted to ask for a review before working on integration level tests.

File: encoderfile/src/builder/image_preprocessor.rs (hunk 14 above)
Comment (javiermtorres) [reply encoderfile/src/builder/image_preprocessor.rs]: encoderfile_config finally, it's clearer after all...

File: encoderfile/src/transport/http/example.md (hunk 19 above)
Comment (javiermtorres) [reply encoderfile/src/transport/http/example.md:19]: BTW it now really works :D

**Comment by javiermtorres** [comment]
> unit tests + formatting, otherwise this looks great!

Thanks a lot @besaleli 🤗 I'll make sure to add more unit tests and fix checks before merging 👍🏼
